src/Psl/DateTime/DateTime.php

```diff
@@ -0,0 +1,552 @@
+    }
+
+    /**
+     * Returns the hours.
+     *
+     * @return int<0, 23>
+     *
+     * @psalm-mutation-free
+     */
+    public function getHours(): int
+    {
+        return $this->hours;
+    }
+
+    /**
+     * Returns the minutes.
+     *
+     * @return int<0, 59>
+     *
+     * @psalm-mutation-free
+     */
+    public function getMinutes(): int
+    {
+        return $this->minutes;
+    }
+
+    /**
+     * Returns the seconds.
+     *
+     * @return int<0, 59>
+     *
+     * @psalm-mutation-free
+     */
+    public function getSeconds(): int
+    {
+        return $this->seconds;
+    }
+
+    /**
+     * Returns the nanoseconds.
+     *
+     * @return int<0, 999999999>
+     *
+     * @psalm-mutation-free
+     */
+    public function getNanoseconds(): int
+    {
+        return $this->nanoseconds;
+    }
+
+    /**
+     * Gets the timezone associated with the date and time.
+     *
+     * @psalm-mutation-free
+     */
+    public function getTimezone(): Timezone
+    {
+        return $this->timezone;
+    }
+
+    /**
+     * Returns a new instance with the specified date.
+     *
+     * @param Month|int<1, 12> $month
+     * @param int<1, 31> $day
+     *
+     * @throws Exception\UnexpectedValueException If any of the provided date components do not align with calendar expectations.
+     *
+     * @psalm-mutation-free
+     */
+    public function withDate(int $year, Month|int $month, int $day): static
+    {
+        return static::fromParts(
+            $this->getTimezone(),
+            $year,
+            $month,
+            $day,
+            $this->getHours(),
+            $this->getMinutes(),
+            $this->getSeconds(),
+            $this->getNanoseconds(),
+        );
+    }
+
+    /**
+     * Returns a new instance with the specified time.
+     *
+     * @param int<0, 23> $hours
+     * @param int<0, 59> $minutes
+     * @param int<0, 59> $seconds
+     * @param int<0, 999999999> $nanoseconds
+     *
+     * @throws Exception\UnexpectedValueException If any of the provided time components do not align with calendar expectations.
+     *
+     * @psalm-mutation-free
+     */
+    public function withTime(int $hours, int $minutes, int $seconds = 0, int $nanoseconds = 0): static
+    {
+        return static::fromParts(
+            $this->getTimezone(),
+            $this->getYear(),
+            $this->getMonth(),
+            $this->getDay(),
+            $hours,
+            $minutes,
+            $seconds,
+            $nanoseconds,
+        );
+    }
+
+    /**
+     * Adds the specified duration to this date-time object, returning a new instance with the added duration.
+     *
+     * @throws Exception\UnderflowException If adding the duration results in an arithmetic underflow.
+     * @throws Exception\OverflowException If adding the duration results in an arithmetic overflow.
+     *
+     * @psalm-mutation-free
+     */
+    public function plus(Duration $duration): static
```

Comment:
previous because `fromTimestamp` was not part of the interface and was specfic to this class, i think we can move those to the trait now, trying to have as many things there as possible